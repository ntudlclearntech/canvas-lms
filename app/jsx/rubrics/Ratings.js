/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Text from '@instructure/ui-elements/lib/components/Text'
import I18n from 'i18n!edit_rubric'

import { ratingShape, tierShape } from './types'

const pointString = (points) =>
  I18n.t('%{points} pts', {
    points: I18n.toNumber(points, { precision : 1 })
  })

const Rating = (props) => {
  const {
    description,
    long_description,
    points,
    onClick,
    classes
  } = props

  return (
    <div
      className={classes}
      onClick={onClick}
      onKeyPress={(e) => e.key === 'Enter' ? onClick() : null}
      role="button"
      tabIndex={0}
    >
      <div className="rating-points">
        <Text size="x-small">
          {pointString(points)}
        </Text>
      </div>
      <div className="rating-description">
        <Text size="x-small" lineHeight="condensed">
          {description}
        </Text>
      </div>
      <Text size="x-small" fontStyle="italic" lineHeight="condensed">
        {long_description}
      </Text>
      <div className='shader'><div className="triangle"/></div>
    </div>
  )
}

const getMasteryLevel = (points, masteryThreshold) => {
  if (points >= masteryThreshold) {
    return 'full'
  } else if (points >= masteryThreshold/2) {
    return 'partial'
  } else {
    return 'none'
  }
}

Rating.propTypes = {
  ...tierShape,
  selected: PropTypes.bool
}
Rating.defaultProps = {
  selected: false,
}

const Ratings = ({ assessing, tiers, points, onPointChange, masteryThreshold }) => {
  const pairs = tiers.map((tier, index) => {
    const next = tiers[index + 1]
    return { current: tier.points, next: next ? next.points : 0 }
  })

  const currentIndex = () => pairs.findIndex(({ current, next }) => (
    points > next && points <= current
  ))
  const selectedIndex = points !== undefined ? currentIndex() : null

  return (
    <div className={classNames("rating-tier-list", { 'react-assessing': assessing })}>
      {
        tiers.map((tier, index) => {
          const selected = selectedIndex === index
          const classes = classNames({
            'rating-tier': true,
            'selected': selected,
          }, selected && getMasteryLevel(points, masteryThreshold))
          return ( // eslint-disable-next-line react/no-array-index-key
            <Rating key={index} onClick={() => onPointChange(tier.points)} classes={classes} {...tier} />
          )
        })
      }
    </div>
  )
}
Ratings.propTypes = {
  ...ratingShape,
  onPointChange: PropTypes.func
}
Ratings.defaultProps = {
  onPointChange: () => { }
}

export default Ratings
