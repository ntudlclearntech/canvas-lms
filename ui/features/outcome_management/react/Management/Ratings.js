/*
 * Copyright (C) 2021 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React, {useCallback} from 'react'
import PropTypes from 'prop-types'
import {IconButton} from '@instructure/ui-buttons'
import {Flex} from '@instructure/ui-flex'
import {IconPlusLine} from '@instructure/ui-icons'
import I18n from 'i18n!OutcomeManagement'
import {View} from '@instructure/ui-view'
import {createRating} from '@canvas/outcomes/react/hooks/useRatings'
import useCanvasContext from '@canvas/outcomes/react/hooks/useCanvasContext'
import ProficiencyRating from '../MasteryScale/ProficiencyRating'

const ratingsShape = PropTypes.shape({
  description: PropTypes.string,
  points: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  mastery: PropTypes.bool
})

const Ratings = ({ratings, onChangeRatings, canManage}) => {
  const {isMobileView} = useCanvasContext()

  const addRow = () => {
    let points = 0.0
    const last = ratings[ratings.length - 1]

    if (last) {
      points = last.points - 1.0
    }

    if (points < 0.0 || Number.isNaN(points)) {
      points = 0.0
    }

    onChangeRatings([...ratings, createRating('', points, null, false)])
  }

  const onRatingFieldChange = (field, value, index) => {
    onChangeRatings(currentRatings => {
      let ratingsCopy = [...currentRatings]

      const newRating = {
        ...ratingsCopy[index],
        [field]: value
      }

      const oldRating = ratingsCopy[index]

      // if mastery was set to true here
      // remove mastery from all
      if (newRating.mastery && !oldRating.mastery) {
        ratingsCopy = ratingsCopy.map(r => ({
          ...r,
          mastery: false
        }))
      }

      ratingsCopy.splice(index, 1, newRating)

      return ratingsCopy
    })
  }

  const handleDelete = useCallback(
    currentIndex => {
      onChangeRatings(currentRatings => {
        const ratingsCopy = [...currentRatings]

        const currentRating = currentRatings[currentIndex]
        ratingsCopy.splice(currentIndex, 1)

        if (currentRating.mastery) {
          const newMasteryIndex = Math.max(currentIndex - 1, 0)
          ratingsCopy[newMasteryIndex] = {
            ...ratingsCopy[newMasteryIndex],
            mastery: true
          }
        }

        return ratingsCopy
      })
    },
    [onChangeRatings]
  )

  const renderBorder = () => (
    <View
      width="100%"
      textAlign="start"
      margin="0 0 small 0"
      as="div"
      borderWidth="none none small none"
    />
  )

  return (
    <>
      <Flex
        width="100%"
        padding={isMobileView ? '0 0 small 0' : '0 small small small'}
        margin="medium none none"
        data-testid="outcome-management-ratings"
      >
        <Flex.Item size={isMobileView ? '25%' : '15%'} padding="0 medium 0 0">
          <div aria-hidden="true" className="header">
            {I18n.t('Mastery')}
          </div>
        </Flex.Item>
        <Flex.Item size={isMobileView ? '75%' : '65%'}>
          <div aria-hidden="true" className="header">
            {I18n.t('Proficiency Rating')}
          </div>
        </Flex.Item>
        {!isMobileView && (
          <Flex.Item size="10%">
            <div aria-hidden="true" className="header">
              {I18n.t('Points')}
            </div>
          </Flex.Item>
        )}
      </Flex>
      {renderBorder()}
      {ratings.map(({key, description, descriptionError, pointsError, mastery, points}, index) => (
        <React.Fragment key={key}>
          <ProficiencyRating
            description={description}
            descriptionError={descriptionError}
            disableDelete={ratings.length === 1}
            mastery={mastery}
            onDelete={() => handleDelete(index)}
            onDescriptionChange={value => onRatingFieldChange('description', value, index)}
            onMasteryChange={() => onRatingFieldChange('mastery', true, index)}
            onPointsChange={value => onRatingFieldChange('points', value, index)}
            points={points?.toString()}
            pointsError={pointsError}
            isMobileView={isMobileView}
            position={index + 1}
            canManage={canManage}
            individualOutcome
          />
          {renderBorder()}
        </React.Fragment>
      ))}
      {canManage && (
        <View textAlign="center" padding="small" as="div">
          <IconButton
            onClick={addRow}
            withBorder={false}
            color="primary"
            size="large"
            shape="circle"
            screenReaderLabel={I18n.t('Add Mastery Level')}
          >
            <IconPlusLine />
          </IconButton>
        </View>
      )}
    </>
  )
}

Ratings.propTypes = {
  ratings: PropTypes.arrayOf(ratingsShape).isRequired,
  canManage: PropTypes.bool,
  onChangeRatings: PropTypes.func
}

Ratings.defaultProps = {
  onChangeRatings: () => {}
}

export default Ratings
