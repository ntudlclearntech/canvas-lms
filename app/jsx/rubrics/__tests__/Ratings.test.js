/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import Ratings from '../Ratings'

describe('The Ratings component', () => {
  const props = {
    tiers: [
      { description: 'Superb', points: 10 },
      { description: 'Meh', long_description: 'More Verbosity', points: 5 },
      { description: 'Subpar', points: 1 }
    ],
    points: 5,
    masteryThreshold: 10,
    useRange: false
  }

  const component = (mods) => shallow(<Ratings {...{ ...props, ...mods }} />)
  it('renders the root component as expected', () => {
    expect(component().debug()).toMatchSnapshot()
  })

  it('renders the Rating sub-components as expected when range rating enabled', () => {
    const useRange = true
    component({ useRange }).find('Rating')
      .forEach((el) => expect(el.shallow().debug()).toMatchSnapshot())
  })

  it('highlights the right rating', () => {
    const ratings = (points, useRange = false) =>
      component({ points, useRange }).find('Rating').map((el) => el.shallow().hasClass('selected'))

    expect(ratings(10)).toEqual([true, false, false])
    expect(ratings(8)).toEqual([false, false, false])
    expect(ratings(8, true)).toEqual([true, false, false])
    expect(ratings(5)).toEqual([false, true, false])
    expect(ratings(3)).toEqual([false, false, false])
    expect(ratings(3, true)).toEqual([false, true, false])
    expect(ratings(1)).toEqual([false, false, true])
    expect(ratings(0, true)).toEqual([false, false, true])
    expect(ratings(undefined)).toEqual([false, false, false])
  })

  it('calls onPointChange when a rating is clicked', () => {
    const onPointChange = sinon.spy()
    const el = component({ onPointChange })

    el.find('Rating').first().prop('onClick').call()
    expect(onPointChange.args[0]).toEqual([10])
  })

  it('uses the right mastery level', () => {
    const mastery = (points, mastery_level) =>
      component({ points }).find('Rating').map((el) => el.shallow().hasClass(mastery_level))
    expect(mastery(10, 'full')).toEqual([true, false, false])
    expect(mastery(5, 'partial')).toEqual([false, true, false])
    expect(mastery(1, 'none')).toEqual([false, false, true])
  })
})
