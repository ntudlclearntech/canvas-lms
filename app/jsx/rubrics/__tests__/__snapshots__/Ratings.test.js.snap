// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 1`] = `
"<div className=\\"rating-tier\\" onClick={[Function]} onKeyPress={[Function]} role=\\"button\\" tabIndex={0}>
  <div className=\\"rating-points\\">
    <Text size=\\"x-small\\" as=\\"span\\" letterSpacing=\\"normal\\">
      10.0 to &gt;5.0 pts
    </Text>
  </div>
  <div className=\\"rating-description\\">
    <Text size=\\"x-small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Superb
    </Text>
  </div>
  <Text size=\\"x-small\\" fontStyle=\\"italic\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\" />
  <div className=\\"shader\\">
    <div className=\\"triangle\\" />
  </div>
</div>"
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 2`] = `
"<div className=\\"rating-tier selected partial\\" onClick={[Function]} onKeyPress={[Function]} role=\\"button\\" tabIndex={0}>
  <div className=\\"rating-points\\">
    <Text size=\\"x-small\\" as=\\"span\\" letterSpacing=\\"normal\\">
      5.0 to &gt;1.0 pts
    </Text>
  </div>
  <div className=\\"rating-description\\">
    <Text size=\\"x-small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Meh
    </Text>
  </div>
  <Text size=\\"x-small\\" fontStyle=\\"italic\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\">
    More Verbosity
  </Text>
  <div className=\\"shader\\">
    <div className=\\"triangle\\" />
  </div>
</div>"
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 3`] = `
"<div className=\\"rating-tier\\" onClick={[Function]} onKeyPress={[Function]} role=\\"button\\" tabIndex={0}>
  <div className=\\"rating-points\\">
    <Text size=\\"x-small\\" as=\\"span\\" letterSpacing=\\"normal\\">
      1.0 to &gt;0.0 pts
    </Text>
  </div>
  <div className=\\"rating-description\\">
    <Text size=\\"x-small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Subpar
    </Text>
  </div>
  <Text size=\\"x-small\\" fontStyle=\\"italic\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\" />
  <div className=\\"shader\\">
    <div className=\\"triangle\\" />
  </div>
</div>"
`;

exports[`The Ratings component renders the root component as expected 1`] = `
"<div className=\\"rating-tier-list\\">
  <Rating onClick={[Function]} classes=\\"rating-tier\\" endOfRangePoints={{...}} description=\\"Superb\\" points={10} selected={false} />
  <Rating onClick={[Function]} classes=\\"rating-tier selected partial\\" endOfRangePoints={{...}} description=\\"Meh\\" long_description=\\"More Verbosity\\" points={5} selected={false} />
  <Rating onClick={[Function]} classes=\\"rating-tier\\" endOfRangePoints={{...}} description=\\"Subpar\\" points={1} selected={false} />
</div>"
`;
